torch/csrc/LazyTensor.cpp

```diff
@@ -0,0 +1,385 @@
+    auto fs = lt->schema();
+    auto s = Symbol::fromQualString(fs.name());
+    auto* node = g->create(s, fs.returns().size());
+    if (dep) {
+      g->setInsertPoint(dep);
+    }
+    node = g->insertNode(node);
+    return node;
+  };
+
+  // Setup output of the whole graph
+  auto* root_node = createNode(lt_, nullptr);
+  torch::jit::Value* output = root_node->outputs()[lt_->outputIndex()];
+  std::vector<const LazyTensorImpl*> inputs;
+
+  // Create frontier for resolving deps
+  // The form is { IValue, Node (that depends on IValue) }
+  std::vector<std::pair<c10::IValue, torch::jit::Node*>> frontier;
+  for (const auto& input : lt_->inputs()) {
+    frontier.emplace_back(std::make_pair(input, root_node));
+  }
+
+  while (frontier.size()) {
+    std::vector<std::pair<c10::IValue, torch::jit::Node*>> next_frontier;
+    for (auto dep_pair : frontier) {
+      auto ival = dep_pair.first;
+      auto dep_node = dep_pair.second;
+
+      // Is the IValue a LazyTensor?
+      if (ival.isTensor() && ival.toTensor().is_lazy()) {
+        auto t = ival.toTensor();
+        auto* lt = static_cast<LazyTensorImpl*>(t.unsafeGetTensorImpl());
+        // Is lt just storing a tensor? This is an input
+        if (lt->isResolved()) {
+          auto* v = g->addInput();
+          inputs.emplace_back(lt);
+          dep_node->addInput(v);
+        } else {
+          auto* node = createNode(lt, dep_node);
+          dep_node->addInput(node->outputs()[lt->outputIndex()]);
+          for (const auto& input : lt->inputs()) {
+            next_frontier.emplace_back(std::make_pair(input, node));
+          }
+        }
+        // It's a non-LazyTensor
+      } else {
+        g->setInsertPoint(dep_node);
+        torch::jit::Value* v;
+        if (ival.isNone()) {
+          v = g->insertNode(g->createNone())->output();
+        } else {
+          v = insertConstant(*g, ival);
+        }
+        dep_node->addInput(v);
+      }
+    }
+
+    frontier = next_frontier;
+  }
+
+  return std::make_pair(output->node(), inputs);
+}
+
+/* Cache bytecode
+ *
+ * We cache the compiled graphs produced by lazy tensors,
+ * acting under the assumption that we will repeatedly call the same
+ * graphs.  This cache is implemented with a simple bytecode used
+ * as the key.  It is a vector containing elements of two types:
+ * 1) schema strings + output indices
+ * 2) ivalues themselves
+ *
+ * We can see each element is isomorphic to the representation of a LazyTensor
+ * and it is straightforward to extend this model to a bytecode representing
+ * the underlying graph using prefix notation.
+ *
+ * Here's an example:
+ *   Construct a straight line repr of the LT
+ *    a = ?.to_lazy()
+ *    b = ?.to_lazy()
+ *    c = a @ b
+ *    d = c * 2
+ *   this becomes
+ *    mul[0], mm[0], 2, a, b
+ *   that is, all unresolved LTs in the graph are "ExprName[index]" and
+ *   resolved LTs are just their corresponding IValues
```

Comment:
Use of prefix notation seems questionable to me, because it unflattens graphs into trees and can, once again, lead to exponential size representations. Wouldn't some sort of hash consing approach be better?